Adds Development Guidelines working document

narbs91 opened the pull request:
### What does it do?
Adds the Developer Guidelines working doc from its current Google Doc form to the project repo.  A more condensed version will be transferred to the CONTRIBUTING file once we refine our ideas in the bigger document.

### Any helpful background information?

As a community we want to be as accessible as possible to any new developers that want to start contributing and not wonder how/where to start.  Having a foundation for best practices and how to be a rockstar Contributor would provide a great resources for anyone to go from 0 to 1 real quick in the Developer DAO

### Any new dependencies? Why were they added?
DEVELOPMENT_GUIDELINES document added to remove Google Docs dependency

### Relevant screenshots/gifs

### Does it close any issues?

Closes #112

with-heart commented on docs/DEVELOPMENT_GUIDELINES.md:
```diff
@@ -0,0 +1,201 @@
+project, be sure to run yarn to install the project's dependencies. Developers
+should ensure that package-lock.json is not committed with their code.
+
+# Commitment Statements
```
I find this whole section pretty weird if I'm being honest. I've never heard of a commitment statement before and a few things jump out at me about them:

- They seem like something you could maybe get away with on a mature team in a mature org, but we aren't even remotely that (yet)
- They set a lot of expectations for a single person that maybe that person doesn't want to care about or doesn't need to care about (see next point)
- Open source works really well because it allows people to focus on what they're good at or are passionate about and learn about things as they go along, and we fill in the gaps by helping each other out, reviewing each other's code, and teaching each other things

There's nothing wrong with having some quality checks in place, but for a lot of these things, that's what code review and good maintainers are for 😁 

We have to trust that contributors are trying their best, we have to trust that the community will help newer contributors who don't yet know what a lot of these things are, and we have to trust that we can foster a culture of effective code reviews and maintainers. And this just feels like us saying "we don't trust you unless you repeat after me" lol

narbs91 replied:
This section was all me and this is something that we have at the company I work for.  I was curious to see if/how it would stick in an environment like this but that's mostly from my lack of open-source experience.  This is probably something that might make sense way down but yeah, just reflecting now, it's a bit premature to have at this point

with-heart replied:
I think you should always lean towards suggesting things regardless of your background! We can only find what works if we talk through all the options ya know?

narbs91 replied:
Cool, cool yeah makes sense to me, that's the only way we'll see what'll stick and what won't haha

with-heart reviewed (commented):
Dumped various thoughts throughout! 

Overall I guess I would sum the thoughts up as: 

I think the best way for us to move forward quickly, foster a healthy community, unleash lots of creativity, and level up as a collective is to avoid defining process stuff until we feel a legitimate pain that needs a process to solve.

We might find that we end up needing to define exactly the processes and expectations laid out in this document that I pushed back on, but I think it's better for us to _experience the problem first_ to make sure we understand what it is we're solving for before we attempt to solve it.

Open source and the DevDAO should imo be FUN first and foremost, and I think too much process defined inorganically can really stifle that

narbs91 commented:
Thanks for all the awesome feedback @with-heart, I'll be interested to hear what the rest of the gang thinks too.  I think at this point of the game, the entire document is probably overkill and too many steps ahead of where we're at currently.  The aim of this PR i think should be to cut down to the essential pieces of getting someone from 0 to 1 in terms of how to start contributing, just having some common ground between all of us.  

The last thing i want is this awesomeness to turn political and hand pointy

narbs91 commented on docs/DEVELOPMENT_GUIDELINES.md:
```diff
@@ -0,0 +1,201 @@
+languages cannot be incorporated and / or cannot be completed at the moment,
+English should be the default fallback language.
+
+TBD - What is the best (automated?) way to ensure correct translation into other
+languages? Using google translate as a source should be discouraged as it often
+results in mis-translations
```
Yeah no way around this one haha

narbs91 commented on docs/DEVELOPMENT_GUIDELINES.md:
```diff
@@ -0,0 +1,201 @@
+Questions are also welcome as issues, which can help with defining better
+documentation.
+
+Proposals that include UI changes should preferably be accompanied by design
+mockups that illustrate how the changes would look like. This facilitates the
+discussion about the proposal and can potentially save development time as well.
+You can copy (akin to forking) the Figma design file from this link:
+https://www.figma.com/file/BBxYJrArJFOpND9c4OBWkY/D_D?node-id=3%3A2 , make your
+proposed changes in your own file, and include a link in your issue.
```
Makes sense to me

narbs91 commented on docs/DEVELOPMENT_GUIDELINES.md:
```diff
@@ -0,0 +1,201 @@
+
+3. Proposed / Solution / How It Can Be Improved / Fixed
+
+4. How Tests Will Be Performed: Manual & Regression / Unit Tests / E2E Tests
```
I guess it comes down to how fundamental we want to get here and the level of assumptions we have around the experience people have around general software development.  It's hard to say where to draw the line on what assumptions we should make (if any).  I think eventually, I'd love for our community to be able to provide enough learning material/resources to help anyone take the step into becoming a developer/designer/etc... regardless of prior experience

with-heart replied:
YES, I would love that too! That makes a lot of sense.

I think in regards to that it's like I'd rather we set expectations for the community as a whole and/or ourselves rather than putting things on individual contributors. 

So like we need a robust test suite and we want to make expanding that suite a habit for contributors, and imo the best way for us to get that is for people who care about the test suite to be active in PRs. Pair with the contributor if they've never tested something before, make helpful suggestions and give thorough critique, share ideas with them, contribute the tests ourselves if it's what's needed, etc. 

If we expect that level of commitment of ourselves to treat contributors with empathy and to view code reviews as the _best_ way for us to collectively improve each other, I think we'll find a significant amount of success!

narbs91 replied:
Totally agree, that type of mentoring will only be a win-win for the whole community

Dhaiwat10 commented:
I agree with @with-heart. Overall, this is good but in some places, it sounds way too restrictive, hostile, _not_ welcoming. Can we make this a bit more friendly? This sounds a wee bit too serious haha

narbs91 commented:
> I agree with @with-heart. Overall, this is good but in some places, it sounds way too restrictive, hostile, _not_ welcoming. Can we make this a bit more friendly? This sounds a wee bit too serious haha.
> 
> I'll try to find some time & suggest edits myself if possible. :)

Yessir, working on a condensed and non-hostile version now haha.  I'm leaving the DEVLEOPMENT_GUIDELINES doc as is and putting the new version inside the CONTRIBUTING.md doc

narbs91 commented:
Heya gang, i've updated the contributing doc with a condensed and more welcoming tone using the original doc as a guide and the comments (so far) to amend.  I've also added a couple of issue templates for feature requests and bug reporting (also based on the content in the original doc).  Please have a look whenever you get a chance.  In regards to the original doc, i think we can leave it be for now and adjust it as we continue to evolve as an org.  Keeping it in the repo will at least make it accessible to everyone in the community.

narbs91 commented:
CC: @codingwithmanny @Dhaiwat10 @with-heart @nagmak

Dhaiwat10 commented on README.md:
```diff
@@ -87,7 +87,7 @@ nvm install; # to install the version in .nvmrc
 ## Contributing
 
-Thanks for showing interest in contributing to DeveloperDAO. Before submitting any changes please review our contributing gudielines in [CONTRIBUTING.md](./CONTRIBUTING.md).
+Thanks for showing interest in contributing to DeveloperDAO. Before submitting any changes please review our contributing gudielines in [CONTRIBUTING.md](./docs/CONTRIBUTING.md).
```
Typo!

```suggestion
Thanks for showing interest in contributing to Developer DAO. Before submitting any changes please review our contributing gudielines in [CONTRIBUTING.md](./docs/CONTRIBUTING.md).
```

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+Thanks for showing interest in contributing to the Developer DAO(D_D)!
```
```suggestion
Thanks for showing interest in contributing to Developer DAO(D_D)!
```

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+Thanks for showing interest in contributing to the Developer DAO(D_D)!
+
+This document will help you figure how to get started contributing as quickly as
+possible. The target audience of this document is literally anybody who wants to
+contribute to the Developer-DAO organization, regardless of skill level or
+technical background, in the manner that fits them the best
```
```suggestion
This document will help you to start contributing to the DAO as quickly as
possible. The target audience of this document is literally anybody who wants to
contribute to Developer DAO, regardless of skill level or
technical background, in the manner that fits them the best.
```

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+contribute to the Developer-DAO organization, regardless of skill level or
+technical background, in the manner that fits them the best
+
+# Contributing your way
+
+As mentioned earlier, if you want to contribute to this fantastic organization
+you should feel empowered to do so in the way that best suites your interests.
```
I feel this is redundant.

```suggestion
```

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+
+## Ways to contribute
+
+We all come from a wide range of experiences, skill-sets and backgrounds;
+however, we're all here because of the opportunity to contribute to pushing web3
+forward and growing together. There's no one way of contributing to this
+organization which is what makes it so awesome. You're an out-of-this-world
+designer? Sweet, feel free to suggest anything to improve our current design.
+Are you passionate about testing? Excellent, you can look to ways we can improve
+our current test suites and ensure developers are including units tests through
+code reviews. Indeed, there are many ways one can contribute, below are the
+current areas most community members are helping out with today:
```
Some re-wording.

```suggestion
We all come from a wide range of experiences, skill-sets, and backgrounds;
however, we're all here because of the opportunity to contribute to pushing web3
forward and growing together. There's no one way of contributing to this
DAO which is what makes it so awesome. You're an out-of-this-world
designer? Sweet, feel free to suggest anything to improve our current design.
Are you passionate about testing? Excellent, you can look to ways we can improve
our current test suites and ensure developers are including units tests through
code reviews. Indeed, there are many ways one can contribute. Below are the areas most community members are helping out with:
```

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+
+## Picking up an issue
+
+Oh yeah, you got this far and are looking to get some skin in the game, very
+wise choice. If your itching to get straight into the action, go to the
+[issues tab](https://github.com/Developer-DAO/developer-dao/issues) and see
+which of the open issues interests you the most. How do you know an issue hasn't
+been picked up you ask? Issues are typically picked up by commenting on the
+given issue that you'd be interested in picking taking on the effort and then
+assigning it to yourself (or having someone do that for you). Once you're
+feature is completed and out in the world, please take the time to close out the
+issue itself so that we can properly say what stuff we've shipped and what's
+actually left to do.
```
No need to manually close the issue. Our issue template does that for us. :)

```suggestion
Oh yeah, you got this far and are looking to get some skin in the game! Very
wise choice. If you are itching to get straight into the action, go to the
[issues tab](https://github.com/Developer-DAO/developer-dao/issues) and see
which of the open issues interest you the most. How do you know an issue hasn't
been picked up you ask? Issues are typically picked up by commenting on the
given issue that you'd be interested in picking taking on the effort and then having them assigned to you by someone.
```

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+issue itself so that we can properly say what stuff we've shipped and what's
+actually left to do.
+
+To start contributing, we ask all contributors to fork a copy of this repo so
+that you don't work off the main branch. Once you have the project forked, you
+would then checkout a branch off your local copy of the project via
+`git checkout -b your_branch_name` per issue you pick up. You can name the
+branch what you want but be sure to give it some context around the issue you're
+working on.
```
```suggestion
To start contributing, we ask all contributors to fork this repo. You should then clone your fork and create a new branch using `git checkout -b your_branch_name` per issue you pick up. You can name the branch what you want but be sure to give it some context around the issue you're working on.
```

narbs91 replied:
I feel like this is more a coding style thing and you don't really need to continuously clone the fork.  As long as people are pulling from main before checkout a new branch they should be fine

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+branch what you want but be sure to give it some context around the issue you're
+working on.
+
+Once you're well on your way through your work, don't be shy/hesitant to post a
+draft PR to have the community take a look at the progress or provide helpful
+insight/tips/suggestions on how you can improve the current implementation. At
+the end of the day the community is here to help one another and to ensure
+people are getting the most out their experience being part of the
+Developer_DAO.
```
```suggestion
Once you're well on your way through your work, don't be shy/hesitant to open a
draft pull request to have the community take a look at the progress or help you with your task. At the end of the day, the community is here to help one another and to ensure
people are getting the most out of their experience being part of Developer DAO.
```

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+
+## Issue creation
+
+GitHub Issues are the primary way this community discusses ideas and plans for
+future work for the website. We have
+[a number of issue forms](https://github.com/Developer-DAO/developer-dao/issues/new/choose)
+to help guide you through the process of sharing a feature request or reporting
+a bug.
```
```suggestion
GitHub Issues are the primary way we discuss ideas and plans for
future work for the website. We have [a number of issue forms](https://github.com/Developer-DAO/developer-dao/issues/new/choose) to help guide you through the process of sharing a feature request or reporting a bug.
```

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+in functionality, better approaches to implementation and so on. Let's aim to be
+mentors to those who want to contribute and give the community the opportunity
+to grow together. There are major learning opportunities to be had with code
+reviews, especially to the more inexperienced among us who are here to learn
```
```suggestion
reviews, especially to the less experienced among us who are here to learn
```

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+most up to date documentation so that future contributors will have the most
+accurate information available to them. If you see an opportunity to improve
+existing documentation or add something that may be missing, open up an issue
+and get writing.
```
```suggestion
and get writing!
```

Dhaiwat10 commented on docs/CONTRIBUTING.md:
```diff
@@ -0,0 +1,119 @@
+
+## Translations
+
+Being part of the internet means our audience reach is the entire globe. As such
+we want our audience to actually be able to access our content if english isn't
+their first language. Our code base uses
+[next-i18next](https://github.com/isaachinman/next-i18next) for translation
+purposes and our actual translations are provided by our awesome community
+members. If you're a whiz in a given language we support (or don't have yet),
+please raise your hand to help out with making our site as multilingual as
+possible.
```
```suggestion
Being part of the internet means we have a global audience. This is why
we want our users to actually be able to access our content if English isn't
their first language. Our code base uses
[next-i18next](https://github.com/isaachinman/next-i18next) for translation
purposes and our actual translations are provided by our awesome community
members. If you're a whiz in a given language we support (or don't have yet),
please raise your hand to help out with making our site as multilingual as
possible!
```

Dhaiwat10 reviewed (changes requested):
Great stuff.

Some grammar, re-wording, etc. :)

codingwithmanny commented on docs/DEVELOPMENT_GUIDELINES.md:
```diff
@@ -0,0 +1,201 @@
+
+# Issues
+
+Issues should be used to propose new ideas and / or a discussion around a newly
+proposed feature. It should be noted that this discussion should pertain to the
+subject and functionality of that repository. If the discussion or feature is
+out of the scope of that project, this issue should be mentioned as so to either
+be put on track or closed because it does not pertain to the scope of the
+current repository. That does not mean that some repositories depend on
+functionality of repositories and could be used at the discretion of the
+community.
```
@with-heart where we you when I first created this doc? lol

with-heart replied:
Oh you know, just drifting around in the metaverse

codingwithmanny commented on docs/DEVELOPMENT_GUIDELINES.md:
```diff
@@ -0,0 +1,201 @@
+
+# Pull Requests (PRs)
+
+A PR should relate to a proposed issue. If a PR does not relate to a specific
+git issue, then the PR could be closed because the feature was not discussed or
+approved beforehand. A developer should never assume that the community
+absolutely needs specific functionality without first discussion and approval
+from the community.
```
@with-heart We've had a few conversations about this, so now we finally get to have it with what I wrote above 😂 .

My only worry, and it has happened, is that there are some devs that start approving PRs that were proposed by a dev that comes out of nowhere. Not everyone has reviewed or discussed whether or not a feature should be revised/improved or not and it gets merged down. We would then either need to figure out technical debt or deal with potential consequences of it affecting the code.

I do see how this can deter someone from contributing, but I can also see the other side where things get merged down without much thought too.

Open to hearing your thoughts on this.

with-heart replied:
> @with-heart We've had a few conversations about this, so now we finally get to have it with what I wrote above 😂 .

🥊 

I have no doubt that we'll eventually run into a situation exactly like what you're describing. I'm almost certain we'll eventually need more process than "just open a PR" for things that involve big picture stuff like design changes or third-party libraries or whatever. 

But if we wait till that happens to figure out what we need, we can make sure the process we come up with is one that makes sense for everyone and is targeted at solving the problem directly vs. a one-size-fits-all process that might actually hurt us in a number of ways.

codingwithmanny replied:
@codingwithmanny 

> I have no doubt that we'll eventually run into a situation exactly like what you're describing. I'm almost certain we'll eventually need more process than "just open a PR" for things that involve big picture stuff like design changes or third-party libraries or whatever.

That's sort of what we're going to do with discourse or equivalent app, talk / propose something, and the proceed as needed. I do see scenarios though that may be overkill for opening an issue, like hotfixes or typos, but for bigger features or things that affect the entire community, I still think it's worth having a conversation, talk about adoption, and even give reasons why or why not do something, versus "Here's my PR with a 1,000 of code that I think is better. I think you should merge it."

> But if we wait till that happens to figure out what we need, we can make sure the process we come up with is one that makes sense for everyone and is targeted at solving the problem directly vs. a one-size-fits-all process that might actually hurt us in a number of ways.

Reading this after my last thoughts do make me think that PRs could be a conversation piece as well. I guess the metaphor I have stuck in my head is, I don't show up to your house unannounced with a new washing machine telling you "Hey you should install this, this is better than your brand and better for you", and I might not have even considered what decisions you made in the first place that made you purchase your existing washing machine.

> I think a good metaphor here is AHA. A process is an abstraction and abstracting things too quickly can lead us to the wrong abstraction. If we instead wait till we've felt the pain a few times, we're more likely to pick the right abstraction.

I would probably revise the metaphor and say that Dodds is writing a blog post to make his point about a particular topic, and inviting you to the conversation, versus sending us a large PR stating "I took the liberty of rewriting all your code to fit with AHA principles instead."

with-heart replied:
I'll just submit an example later so you can see what I mean 😉

codingwithmanny commented on docs/DEVELOPMENT_GUIDELINES.md:
```diff
@@ -0,0 +1,201 @@
+
+# Priority
+
+At the discretion of the community, features should roll out (get merged to
+main), in any priority the community sees fit. As a flexible base, if an issue
+is defined before another, that issue should take priority.
```
@with-heart got it. I think this was to avoid merge conflicts and forcing other developers to fix their code based on something they didn't work on.

What would you suggest as a possible solution?

with-heart replied:
Rebasing is a part of life 🤷 jk but only sort of

imo the best tool we have to handle this is having a good squad of maintainers. I frequently try to merge things in an order that results in less pain for others, but there's no way I could put my decision-making process for that into words and it's also just unavoidable sometimes.

codingwithmanny commented on docs/DEVELOPMENT_GUIDELINES.md:
```diff
@@ -0,0 +1,201 @@
+
+# Issue Abandonment / Stale Issues / Stale PRs
+
+TBD - A means to figure out if a developer cannot complete a specific issue
+within a specific time and have it re-assigned to another volunteer.
```
@with-heart got it. Yeah my thinking behind this is that certain features can become irrelevant or other devs are relying on specific features.

Take for instance the adding params for ids (not saying this was delayed, just took a bit longer to figure out) and the new website design. If the design gets finished before hand, then these features will likely mean less.

It's just also to identify what a threshold is for a stale issue / pr, aka a dev hasn't responded in a while, can it be re-assigned or opened up for another volunteer to take it over.

codingwithmanny commented on docs/DEVELOPMENT_GUIDELINES.md:
```diff
@@ -0,0 +1,201 @@
+
+## Code Formatting
+
+Please use prettier in your new repositories whenever possible and relevant to
+auto-format your code.
```
@with-heart might be good for us to establish the `devdao-linting-formatting-project-base-js`

narbs91 commented:
[build-ci]

narbs91 commented:
CC: @with-heart @Dhaiwat10 @codingwithmanny When you have a chance have a look over the revisions for the CONTRIBUTING.md file and the templates i've added and we can get this PR merged.  As discussed in the meeting today, I've added the DRAFT prefix to the DEVELOPER_GUIDE so that we can keep updating the doc as we go as well as giving full visibility to the community.

Dhaiwat10 commented on .github/ISSUE_TEMPLATE/bug-report.yml:
```diff
@@ -0,0 +1,52 @@
+      label: What happened?
+      description: Also tell us, what did you expect to happen?
+      placeholder: Tell us what you see!
+      value: "A bug happened!"
+    validations:
+      required: true
+  - type: textarea
+    id: repo-steps
+    attributes:
+      label: How to reproduce the bug?
+      description: A brief description of how to reproduce the bug
+      placeholder: Tell us how to repo the bug!
+      value: "A bug happened!"
+    validations:
+      required: true
+  - type: dropdown
+    id: browsers
+    attributes:
+      label: What browsers are you seeing the problem on?
+      multiple: true
+      options:
+        - Firefox
+        - Chrome
+        - Safari
+        - Microsoft Edge
+  - type: dropdown
+    id: devices
+    attributes:
+      label: What device type(s) did you see the problem on?
+      multiple: true
+      options:
+        - Desktop
+        - Mobile
+        - Tablet
+  - type: textarea
+    id: logs
+    attributes:
+      label: Relevant log output
+      description: Please copy and paste any relevant log output. This will be automatically formatted into code, so no need for backticks.
+      render: shell
```
Adding a place where we ask for a screenshot / a recording / any additional context would be handy.

narbs91 replied:
Good call, done

Dhaiwat10 reviewed (changes requested):
We can def change a lot of things in the development guidelines, but it's good enough for a first draft! :D Let's ship this & iterate on it.

narbs91 commented:
Thanks for the look over @Dhaiwat10 ! @codingwithmanny / @with-heart if you get a chance this weekend or Monday, have a look over so we can get this PR out 💪

codingwithmanny reviewed (approved):
LGTM
